File: src/events.rs
use crate::{
    battery::ChargingStatus,
    system_state::{State, SystemState, SystemStateError},
};
use std::{
    fmt, io,
    os::unix::io::AsRawFd,
    time::{Duration, Instant},
};
use udev::MonitorBuilder;

#[derive(Clone)]
pub enum Event {
    PowerInPlug,
    PowerUnPlug,

    PeriodicCheck,
    LowBattery,
    HighCpuLoad,
    LowCpuLoad,

    Unknown,
    Error(String),
}

impl Event {
    const HIGH_CPU_LOAD: f64 = 35.0;
    const LOW_CPU_LOAD: f64 = 30.0;

    fn state_transition(self: &Event, system_state: &SystemState) {
        let old_state = *system_state.state.borrow();

        let new_state = match old_state {
            State::Performance => match self {
                Event::PowerInPlug => State::Performance,
                Event::PowerUnPlug | Event::LowBattery => State::Powersave,
                Event::HighCpuLoad | Event::LowCpuLoad => State::Performance,
                _ => old_state,
            },
            State::Balanced => match self {
                Event::PowerInPlug => State::Performance,
                Event::PowerUnPlug | Event::LowBattery => State::Powersave,
                Event::HighCpuLoad | Event::LowCpuLoad => State::Performance,
                _ => old_state,
            },
            State::Powersave => match self {
                Event::PowerInPlug => State::Performance,
                Event::PowerUnPlug | Event::LowBattery => State::Powersave,
                _ => old_state,
            },
        };

        *system_state.state.borrow_mut() = new_state;
    }

    fn periodic_check(system_state: &SystemState, cpu_load: f64) -> Result<Event, SystemStateError> {
        let low_battery = system_state.battery_states.read_battery_capacity()? <= 20;
        if low_battery {
            return Ok(Event::LowBattery);
        }

        let charging_status = system_state.battery_states.read_charging_status()?;
        let discharging = charging_status == ChargingStatus::DisCharging;

        let boost = system_state.cpu_states.read_cpu_boost()?;
        let high_cpu_load = cpu_load >= Event::HIGH_CPU_LOAD;
        let low_cpu_load = cpu_load < Event::LOW_CPU_LOAD;

        if high_cpu_load && !discharging && !boost {
            return Ok(Event::HighCpuLoad);
        } else if low_cpu_load && !discharging && boost {
            return Ok(Event::LowCpuLoad);
        }

        if discharging {
            Ok(Event::PowerUnPlug)
        } else {
            Ok(Event::PowerInPlug)
        }
    }

    pub fn handle_event(self: &Event, system_state: &SystemState) -> Result<(), SystemStateError> {
        let cpu_load = system_state.cpu_states.read_cpu_load()?;

        let event = Self::periodic_check(system_state, cpu_load).unwrap_or(self.clone());

        let old_state = *system_state.state.borrow();
        event.state_transition(system_state);
        let new_state = *system_state.state.borrow();

        // in its own branch because cpu boost may change depending on cpu load
        if new_state == State::Performance {
            system_state.set_performance_mode(cpu_load >= Event::HIGH_CPU_LOAD)?;
            return Ok(());
        }

        if old_state != new_state {
            match new_state {
                State::Powersave => system_state.set_powersave_mode()?,
                State::Balanced => system_state.set_balanced_mode()?,
                State::Performance => unreachable!(), // Handled by branch above
            }
        }

        Ok(())
    }
}

impl fmt::Display for Event {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Event::PowerInPlug => write!(f, "power plugged in"),
            Event::PowerUnPlug => write!(f, "power un plugged"),

            Event::PeriodicCheck => write!(f, "periodic check"),

            Event::LowBattery => write!(f, "low battery"),
            Event::LowCpuLoad => write!(f, "low cpu load"),
            Event::HighCpuLoad => write!(f, "high cpu load"),

            Event::Unknown => write!(f, "unknown event occured"),
            Event::Error(err) => write!(f, "an error occured: {}", err),
        }
    }
}

pub struct EventPoller {
    socket: udev::MonitorSocket,
    last_periodic_check: Instant,
    periodic_interval: Duration,
}

impl EventPoller {
    pub fn new(interval_duration_s: u8) -> io::Result<Self> {
        let socket = MonitorBuilder::new()?
            .match_subsystem("power_supply")?
            .listen()?;

        Ok(Self {
            socket,
            last_periodic_check: Instant::now(),
            periodic_interval: Duration::from_secs(interval_duration_s.into()),
        })
    }

    pub fn poll_events(&mut self) -> Event {
        let elapsed = self.last_periodic_check.elapsed();
        let timeout_ms = if elapsed >= self.periodic_interval {
            0
        } else {
            (self.periodic_interval - elapsed).as_millis() as i32
        };

        let mut fds = [libc::pollfd {
            fd: self.socket.as_raw_fd(),
            events: libc::POLLIN,
            revents: 0,
        }];

        let result = unsafe { libc::poll(fds.as_mut_ptr(), 1, timeout_ms) };

        if result < 0 {
            return Event::Error(io::Error::last_os_error().to_string());
        }

        if self.last_periodic_check.elapsed() >= self.periodic_interval {
            self.last_periodic_check = Instant::now();
            return Event::PeriodicCheck;
        }

        for event in self.socket.iter() {
            if event.event_type() == udev::EventType::Change {
                if let Some(name) = event.property_value("POWER_SUPPLY_NAME") {
                    let name_str = name.to_str().unwrap_or("");

                    if name_str == "ACAD"
                        || name_str == "AC"
                        || name_str == "ADP1"
                        || name_str == "AC0"
                    {
                        if let Some(online) = event.property_value("POWER_SUPPLY_ONLINE") {
                            let online_str = online.to_str().unwrap_or("");

                            match online_str {
                                "1" => return Event::PowerInPlug,
                                "0" => return Event::PowerUnPlug,
                                _ => return Event::Unknown,
                            }
                        }
                    }
                }
            }
        }

        Event::Unknown
    }
}

File: src/events.zig
const std = @import("std");
const c = @cImport({
    @cInclude("libudev.h");
});
const l_system_state = @import("system_state.zig");

const mem = std.mem;
const SystemState = l_system_state.SystemState;
const State = l_system_state.State;

pub const Event = enum {
    PowerInPlug,
    PowerUnPlug,

    PeriodicCheck,
    LowBattery,
    HighCpuLoad,
    LowCpuLoad,

    Unknown,
};

pub const EventPoller = struct {
    udev: *c.udev,
    monitor: *c.udev_monitor,
    last_periodic_check: std.time.Instant,
    periodic_interval_ns: u64,

    const HIGH_CPU_LOAD = 45.0;
    const LOW_CPU_LOAD = 40.0;

    pub fn init(interval_s: u64) !@This() {
        const udev = c.udev_new() orelse return error.UdevInitFailed;
        errdefer _ = c.udev_unref(udev);
        const monitor = c.udev_monitor_new_from_netlink(udev, "udev")
            orelse return error.MonitorCreationFailed;
        errdefer _ = c.udev_monitor_unref(monitor);

        _ = c.udev_monitor_filter_add_match_subsystem_devtype(monitor, "power_supply", null);
        _ = c.udev_monitor_enable_receiving(monitor);

        return .{
            .udev = udev,
            .monitor = monitor,
            .last_periodic_check = try std.time.Instant.now(),
            .periodic_interval_ns = interval_s * std.time.ns_per_s,
        };
    }

    pub fn deinit(self: *@This()) void {
        _ = c.udev_monitor_unref(self.monitor);
        _ = c.udev_unref(self.udev);
    }

    pub fn poll_events(self: *@This()) !Event {
        const now = try std.time.Instant.now();
        const elapsed = now.since(self.last_periodic_check);

        if (elapsed >= self.periodic_interval_ns) {
            self.last_periodic_check = now;
            return Event.PeriodicCheck;
        }

        const diff = self.periodic_interval_ns - elapsed;
        const timeout_ms: i32 = @intCast(diff / std.time.ns_per_ms);

        const fd = c.udev_monitor_get_fd(self.monitor);
        var fds = [1]std.posix.pollfd{.{
            .fd = fd,
            .events = std.posix.POLL.IN,
            .revents = 0,
        }};

        const poll_res = try std.posix.poll(&fds, timeout_ms);

        if (poll_res > 0) {
            const dev = c.udev_monitor_receive_device(self.monitor);
            if (dev != null) {
                defer _ = c.udev_device_unref(dev);

                const action = c.udev_device_get_action(dev);
                if (action == null or !mem.eql(u8, mem.span(action), "change")) {
                    return Event.Unknown;
                }

                const name = c.udev_device_get_property_value(dev, "POWER_SUPPLY_NAME");
                if (name) |n| {
                    const name_str = mem.span(n);
                    if (is_ac_adapter(name_str)) {
                        const online = c.udev_device_get_property_value(dev, "POWER_SUPPLY_ONLINE");
                        if (online) |o| {
                            const online_str = mem.span(o);
                            if (mem.eql(u8, online_str, "1")) return Event.PowerInPlug;
                            if (mem.eql(u8, online_str, "0")) return Event.PowerUnPlug;
                        }
                    }
                }
            }
        }

        return Event.Unknown;
    }

    fn is_ac_adapter(name: []const u8) bool {
        const adapters = [_][]const u8{ "ACAD", "AC", "ADP1", "AC0" };
        for (adapters) |a| {
            if (mem.eql(u8, name, a)) return true;
        }
        return false;
    }

    fn state_transition(event: Event, system_state: *SystemState) void {
        const old_state = system_state.state;
        system_state.state = switch (old_state) {
            .Performance => switch (event) {
                .PowerInPlug => .Performance,
                .PowerUnPlug => .Powersave,
                .LowBattery => .Powersave,

                .HighCpuLoad => .Performance, // turn boost on
                .LowCpuLoad => .Performance, // turn boost off when load drops

                else => old_state,
            },
            .Balanced => switch (event) {
                .PowerInPlug => .Performance,
                .PowerUnPlug => .Powersave,
                .LowBattery => .Powersave,

                .HighCpuLoad => .Performance, // turn boost on
                .LowCpuLoad => .Performance,

                else => old_state,
            },
            .Powersave => switch (event) {
                .PowerInPlug => .Performance,
                .PowerUnPlug => .Powersave,
                .LowBattery => .Powersave,

                else => old_state,
            },
        };
    }

    fn periodic_check(system_state: *SystemState, cpu_load: f64) !Event {
        const low_battery = try system_state.battery_states.read_battery_capacity() <= 20;
        if (low_battery) return .LowBattery;

        const charging_status = try system_state.battery_states.read_charging_status();
        const discharging = charging_status == .DisCharging;

        const boost = try system_state.cpu_states.read_cpu_boost();
        const high_cpu_load = cpu_load >= HIGH_CPU_LOAD;
        const low_cpu_load = cpu_load < LOW_CPU_LOAD;

        if (high_cpu_load and !discharging and !boost) {
            return .HighCpuLoad;
        } else if (low_cpu_load and !discharging and boost) {
            return .LowCpuLoad;
        }

        if (discharging) return .PowerUnPlug;
        if (!discharging) return .PowerInPlug;

        return .Unknown;
    }

    pub fn handle_event(i_event: Event, system_state: *SystemState) !void {
        const cpu_load = try system_state.cpu_states.read_cpu_load();
        const event = EventPoller.periodic_check(system_state, cpu_load) catch i_event;

        const old_state = system_state.state;
        EventPoller.state_transition(event, system_state);
        const new_state = system_state.state;

        // in its own branch because cpu boost may change depending on cpu load
        if (new_state == .Performance) {
            try system_state.set_performance_mode(cpu_load >= HIGH_CPU_LOAD);
            return;
        }

        if (old_state != new_state) {
            switch (new_state) {
                .Powersave => try system_state.set_powersave_mode(),
                .Balanced => try system_state.set_balanced_mode(),
                .Performance => unreachable,
            }
        }
    }
};
